### PR TITLE
CodeGen: Eliminate dynamic relocations in the register superclass tables.

### PR DESCRIPTION
**Pull request by resistor**

### CODE REVIEW

File: llvm/include/llvm/CodeGen/TargetRegisterInfo.h
Comment (s-barannikov) [inline llvm/include/llvm/CodeGen/TargetRegisterInfo.h:49]: Ideally this would be a full-blown class so as to not expose implementation details to users of `getSuperClasses` (which should return `iterator_range`).
See e.g. `MCRegisterInfo::superregs`.

File: llvm/include/llvm/CodeGen/TargetRegisterInfo.h
Comment (s-barannikov) [inline llvm/include/llvm/CodeGen/TargetRegisterInfo.h:54]: It should return `const TargetRegisterClass &`

Comment (s-barannikov) [reply llvm/include/llvm/CodeGen/TargetRegisterInfo.h:54]: Ah I see it won't always work, some users need the ID, not TargetRegisterClass.
Never mind.

Comment (resistor) [reply llvm/include/llvm/CodeGen/TargetRegisterInfo.h:54]: That's not an obstacle. The bigger problem is that you need a TargetRegisterInfo to do the mapping from IDs back to TargetRegisterClass's, which we can't easily get from inside the iterator.

Comment (s-barannikov) [reply llvm/include/llvm/CodeGen/TargetRegisterInfo.h:54]: One way to do it is to move `superregclasses` to `TargetRegisterInfo` and pass the iterated class as an argument. (This is how similar `MCRegisterInfo` methods are implemented.)
Alternatively, pass `const TargetRegisterInfo &` to `superclasses` and then into the iterator constructor. Not sure if this is worse or better.

That is, it is possible, but since the main goal of this PR is to reduce the number of relocations I think it could be done separately.

**Review by s-barannikov — APPROVED** [review]
LGTM but please wait for @jayfoad review.

**Comment by chandlerc** [comment]
Just to give some motivation, this PR (after #118734 merges) will remove another 13% of non-vtable dynamic relocations from the `clang` binary -- that's over 10k dynamic relocations avoided.

These are also hard to see as no *one* table is terribly large, but there are a large number of them.

There are also more opportunities remaining here, notably the dynamic relocations required in some of the `.*SysRegList` arrays and the `<target>MCRegisterClasses` arrays.

Also, could the same change done here for superclass tables be done for subreg tables? That's the other one that seems to show up as a lot of separate tables, although not as much as superclasses.

**Comment by resistor** [comment]
> Also, could the same change done here for superclass tables be done for subreg tables? That's the other one that seems to show up as a lot of separate tables, although not as much as superclasses.

I'll take a look at the others once this one is in. I expect some of them to be straight forward, but the ID -> TargetRegisterClass table will be tricky.

File: llvm/utils/TableGen/RegisterInfoEmitter.cpp
Comment (jayfoad) [inline llvm/utils/TableGen/RegisterInfoEmitter.cpp:1289]: Could use uint16_t to save space? All in-tree targets have well under 1000 classes currently.

File: llvm/lib/Target/X86/X86RegisterInfo.cpp
Comment (jayfoad) [inline llvm/lib/Target/X86/X86RegisterInfo.cpp:176]: Are you allowed to increment an invalid iterator?

**Review by jayfoad — COMMENTED** [review]
Looks OK to me.

I find the mixture of checking `isValid` vs checking for an explicit `~0U` pretty ugly, but maybe that's inevitable for a sentinel-terminated list.

File: llvm/lib/Target/X86/X86RegisterInfo.cpp
Comment (resistor) [reply llvm/lib/Target/X86/X86RegisterInfo.cpp:176]: In practice it won't break anything here, since the iterator is just a pointer. This whole code path is somewhat awkward, but I don't want to mess with it too much without understanding the underlying logic.

File: llvm/utils/TableGen/RegisterInfoEmitter.cpp
Comment (resistor) [reply llvm/utils/TableGen/RegisterInfoEmitter.cpp:1289]: Probably, but I'd prefer not to make such a functional change mixed in with this PR.

File: llvm/include/llvm/CodeGen/TargetRegisterInfo.h
Comment (resistor) [reply llvm/include/llvm/CodeGen/TargetRegisterInfo.h]: Done

File: llvm/include/llvm/CodeGen/TargetRegisterInfo.h
Comment (resistor) [reply llvm/include/llvm/CodeGen/TargetRegisterInfo.h]: Done

**Comment by resistor** [comment]
> I find the mixture of checking `isValid` vs checking for an explicit `~0U` pretty ugly, but maybe that's inevitable for a sentinel-terminated list.

I figured out a way to get rid of this, by storing the size of the list. PTAL

**Review by jayfoad — COMMENTED** [review]
No specific objection to this patch, but it seems like it could still be cleaned up. If we're storing the size of the list then why do we also need a sentinel value? I'd also be inclined to make `superclasses()` return an `ArrayRef` instead of `iterator_range` (but maybe that's just my own personal prejudice against unnecessary iterators).

File: llvm/include/llvm/CodeGen/TargetRegisterInfo.h
Comment (resistor) [reply llvm/include/llvm/CodeGen/TargetRegisterInfo.h]: Done

**Comment by resistor** [comment]
> No specific objection to this patch, but it seems like it could still be cleaned up. If we're storing the size of the list then why do we also need a sentinel value? I'd also be inclined to make `superclasses()` return an `ArrayRef` instead of `iterator_range` (but maybe that's just my own personal prejudice against unnecessary iterators).

Done.

File: llvm/lib/Target/Hexagon/HexagonRegisterInfo.cpp
Comment (jayfoad) [inline llvm/lib/Target/Hexagon/HexagonRegisterInfo.cpp:435]: I don't see how this would work now. Check `superclasses().empty()` instead?

**Review by jayfoad — COMMENTED** [review]
Thanks!

File: llvm/lib/Target/Hexagon/HexagonRegisterInfo.cpp
Comment (resistor) [reply llvm/lib/Target/Hexagon/HexagonRegisterInfo.cpp]: Done. Seems unfortunate that no Hexxagon tests expose this...

**Comment by vitalybuka** [comment]
Probably from this patch https://lab.llvm.org/buildbot/#/builders/25/builds/4805 ?

**Comment by resistor** [comment]
This other patch definitely caused an ASAN failure and I just reverted it: https://github.com/llvm/llvm-project/pull/119138

**Comment by resistor** [comment]
Looking at the error logs, it definitely appears to be https://github.com/llvm/llvm-project/pull/119138, not this one

**Comment by vitalybuka** [comment]
I see those as well

```
/tmp/bot/llvm-project/llvm/lib/Target/X86/X86RegisterInfo.cpp:177:5: runtime error: applying non-zero offset 4 to null pointer
    #0 0x555b8c2b4bd2 in llvm::X86RegisterInfo::getLargestLegalSuperClass(llvm::TargetRegisterClass const*, llvm::MachineFunction const&) const /tmp/bot/llvm-project/llvm/lib/Target/X86/X86RegisterInfo.cpp:177:5
    #1 0x555b8da1d57d in llvm::RegisterClassInfo::compute(llvm::TargetRegisterClass const*) const /tmp/bot/llvm-project/llvm/lib/CodeGen/RegisterClassInfo.cpp:181:16
    #2 0x555b8da1e63b in llvm::RegisterClassInfo::computePSetLimit(unsigned int) const /tmp/bot/llvm-project/llvm/lib/CodeGen/RegisterClassInfo.cpp:223:3
    #3 0x555b8d5f963c in llvm::RegisterClassInfo::getRegPressureSetLimit(unsigned int) const /tmp/bot/llvm-project/llvm/include/llvm/CodeGen/RegisterClassInfo.h:148:25
    #4 0x555b8d5f61ef in llvm::ScheduleDAGMILive::initRegPressure() /tmp/bot/llvm-project/llvm/lib/CodeGen/MachineScheduler.cpp:1312:36
    #5 0x555b8d5fb60c in llvm::ScheduleDAGMILive::schedule() /tmp/bot/llvm-project/llvm/lib/CodeGen/MachineScheduler.cpp:1459:3
    #6 0x555b8d629ab4 in (anonymous namespace)::MachineSchedulerBase::scheduleRegions(llvm::ScheduleDAGInstrs&, bool) /tmp/bot/llvm-project/llvm/lib/CodeGen/MachineScheduler.cpp:650:17
    #7 0x555b8d627618 in (anonymous namespace)::MachineScheduler::runOnMachineFunction(llvm::MachineFunction&) /tmp/bot/llvm-project/llvm/lib/CodeGen/MachineScheduler.cpp:463:3
    #8 0x555b8d430881 in llvm::MachineFunctionPass::runOnFunction(llvm::Function&) /tmp/bot/llvm-project/llvm/lib/CodeGen/MachineFunctionPass.cpp:94:13
    #9 0x555b8e36b1c3 in llvm::FPPassManager::runOnFunction(llvm::Function&) /tmp/bot/llvm-project/llvm/lib/IR/LegacyPassManager.cpp:1406:27
    #10 0x555b8e38227e in llvm::FPPassManager::runOnModule(llvm::Module&) /tmp/bot/llvm-project/llvm/lib/IR/LegacyPassManager.cpp:1452:16
    #11 0x555b8e36cdba in runOnModule /tmp/bot/llvm-project/llvm/lib/IR/LegacyPassManager.cpp:1521:27
    #12 0x555b8e36cdba in llvm::legacy::PassManagerImpl::run(llvm::Module&) /tmp/bot/llvm-project/llvm/lib/IR/LegacyPassManager.cpp:539:44
    #13 0x555b90a72efb in RunCodegenPipeline /tmp/bot/llvm-project/clang/lib/CodeGen/BackendUtil.cpp:1197:19
    #14 0x555b90a72efb in EmitAssembly /tmp/bot/llvm-project/clang/lib/CodeGen/BackendUtil.cpp:1220:3
    #15 0x555b90a72efb in clang::EmitBackendOutput(clang::DiagnosticsEngine&, clang::HeaderSearchOptions const&, clang::CodeGenOptions const&, clang::TargetOptions const&, clang::LangOptions const&, llvm::StringRef, llvm::Module*, clang::BackendAction, llvm::IntrusiveRefCntPtr<llvm::vfs::FileSystem>, std::__1::unique_ptr<llvm::raw_pwrite_stream, std::__1::default_delete<llvm::raw_pwrite_stream>>, clang::BackendConsumer*) /tmp/bot/llvm-project/clang/lib/CodeGen/BackendUtil.cpp:1383:13
    #16 0x555b90adb00e in clang::CodeGenAction::ExecuteAction() /tmp/bot/llvm-project/clang/lib/CodeGen/CodeGenAction.cpp:1217:3
    #17 0x555b9149559c in clang::FrontendAction::Execute() /tmp/bot/llvm-project/clang/lib/Frontend/FrontendAction.cpp:1077:8
    #18 0x555b9138a94d in clang::CompilerInstance::ExecuteAction(clang::FrontendAction&) /tmp/bot/llvm-project/clang/lib/Frontend/CompilerInstance.cpp:1062:33
    #19 0x555b9175ef91 in clang::ExecuteCompilerInvocation(clang::CompilerInstance*) /tmp/bot/llvm-project/clang/lib/FrontendTool/ExecuteCompilerInvocation.cpp:296:25
    #20 0x555b886f825f in cc1_main(llvm::ArrayRef<char const*>, char const*, void*) /tmp/bot/llvm-project/clang/tools/driver/cc1_main.cpp:286:15
    #21 0x555b886ed85b in ExecuteCC1Tool(llvm::SmallVectorImpl<char const*>&, llvm::ToolContext const&) /tmp/bot/llvm-project/clang/tools/driver/driver.cpp:218:12
    #22 0x555b886ebaa4 in clang_main(int, char**, llvm::ToolContext const&) /tmp/bot/llvm-project/clang/tools/driver/driver.cpp:259:12
```

**Comment by resistor** [comment]
Ok, that's a different error that does come from this commit. Reverting.

**Comment by resistor** [comment]
Reverted in e940353fd2ac

**Pull request merged**
